snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -433,40 +433,15 @@ def _record_manifest_and_source_snapcraft_yaml(self):
                 yaml_utils.dump(annotated_snapcraft, stream=manifest_file)
 
     def write_snap_directory(self) -> None:
-        assets_dir = self._project_config.project._get_snapcraft_assets_dir()
-        # First migrate the snap directory. It will overwrite any conflicting
-        # files.
-        for root, directories, files in os.walk(assets_dir):
-            if root == assets_dir:
-                with contextlib.suppress(ValueError):
-                    directories.remove(".snapcraft")
-                with contextlib.suppress(ValueError):
-                    # The objective of local inside the snap dir is to
-                    # use it as an additional source so there is little
-                    # sense in copying it in here.
-                    directories.remove("local")
-                with contextlib.suppress(ValueError):
-                    # The snapcraft.yaml is migrated later
-                    files.remove("snapcraft.yaml")
-
-            for directory in directories:
-                source = os.path.join(root, directory)
-                destination = os.path.join(self._prime_dir, "snap", directory)
-                file_utils.create_similar_directory(source, destination)
-
-            for file_path in files:
-                source = os.path.join(root, file_path)
-                subtree = os.path.relpath(root, assets_dir)
-                destination = os.path.join(self._prime_dir, "snap", subtree, file_path)
-                with contextlib.suppress(FileNotFoundError):
-                    os.remove(destination)
-                file_utils.link_or_copy(source, destination)
-
-        # Now copy the assets contained within the snap directory directly into
-        # meta.
-        for origin in ["gui", "hooks"]:
-            src_dir = os.path.join(assets_dir, origin)
-            dst_dir = os.path.join(self.meta_dir, origin)
+        snap_assets_dir = self._project_config.project._get_snapcraft_assets_dir()
+        prime_snap_dir = os.path.join(self._prime_dir, "snap")
+
+        snap_dir_iter = itertools.product([prime_snap_dir], ["hooks"])
+        meta_dir_iter = itertools.product([self._meta_dir], ["hooks", "gui"])
+
+        for origin in itertools.chain(snap_dir_iter, meta_dir_iter):
```

Comment:
This is a bit convoluted way to set src_dir and dst_dir, I hope we still understand the logic in the future.